chain/syncer_integration_test.go

```diff
@@ -0,0 +1,617 @@
+	}
+	return initSyncTest(t, con, initGenesisWrapper, cst, bs, r, dstP, chain.Syncing)
+}
+
+func initSyncTest(t *testing.T, con consensus.Protocol, genFunc func(cst *hamt.CborIpldStore, bs bstore.Blockstore) (*types.Block, error), cst *hamt.CborIpldStore, bs bstore.Blockstore, r repo.Repo, dstP *SyncerTestParams, syncMode chain.SyncMode) (*chain.Syncer, *chain.Store, repo.Repo, *th.TestFetcher) {
+	ctx := context.Background()
+
+	calcGenBlk, err := genFunc(cst, bs) // flushes state
+	require.NoError(t, err)
+	calcGenBlk.StateRoot = dstP.genStateRoot
+	chainDS := r.ChainDatastore()
+	chainStore := chain.NewStore(chainDS, cst, calcGenBlk.Cid())
+
+	fetcher := th.NewTestFetcher()
+	syncer := chain.NewSyncer(con, chainStore, fetcher, syncMode) // note we use same cst for on and offline for tests
+
+	// Initialize stores to contain dstP.genesis block and state
+	calcGenTS := th.RequireNewTipSet(t, calcGenBlk)
+
+	genTsas := &chain.TipSetAndState{
+		TipSet:          calcGenTS,
+		TipSetStateRoot: dstP.genStateRoot,
+	}
+	require.NoError(t, chainStore.PutTipSetAndState(ctx, genTsas))
+	err = chainStore.SetHead(ctx, calcGenTS) // Initialize chainStore store with correct dstP.genesis
+	require.NoError(t, err)
+	requireHead(t, chainStore, calcGenTS)
+	requireTsAdded(t, chainStore, calcGenTS)
+
+	return syncer, chainStore, r, fetcher
+}
+
+func containsTipSet(tsasSlice []*chain.TipSetAndState, ts types.TipSet) bool {
+	for _, tsas := range tsasSlice {
+		if tsas.TipSet.String() == ts.String() { //bingo
+			return true
+		}
+	}
+	return false
+}
+
+type requireTsAddedChainStore interface {
+	GetTipSet(types.TipSetKey) (types.TipSet, error)
+	GetTipSetAndStatesByParentsAndHeight(string, uint64) ([]*chain.TipSetAndState, error)
+}
+
+func requireTsAdded(t *testing.T, chain requireTsAddedChainStore, ts types.TipSet) {
+	h, err := ts.Height()
+	require.NoError(t, err)
+	// Tip Index correctly updated
+	gotTs, err := chain.GetTipSet(ts.Key())
+	require.NoError(t, err)
+	require.Equal(t, ts, gotTs)
+	parent, err := ts.Parents()
+	require.NoError(t, err)
+	childTsasSlice, err := chain.GetTipSetAndStatesByParentsAndHeight(parent.String(), h)
+	require.NoError(t, err)
+	require.True(t, containsTipSet(childTsasSlice, ts))
+}
+
+func requireHead(t *testing.T, chain HeadAndTipsetGetter, head types.TipSet) {
+	require.Equal(t, head, requireHeadTipset(t, chain))
+}
+
+func assertHead(t *testing.T, chain HeadAndTipsetGetter, head types.TipSet) {
+	headTipSet, err := chain.GetTipSet(chain.GetHead())
+	assert.NoError(t, err)
+	assert.Equal(t, head, headTipSet)
+}
+
+func requirePutBlocks(_ *testing.T, f *th.TestFetcher, blocks ...*types.Block) types.TipSetKey {
+	var cids []cid.Cid
+	for _, block := range blocks {
+		c := block.Cid()
+		cids = append(cids, c)
+	}
+	f.AddSourceBlocks(blocks...)
+	return types.NewTipSetKey(cids...)
+}
+
+// Syncer is capable of recovering from a fork reorg after Load.
+func TestLoadFork(t *testing.T) {
```

Comment:
@ZenGround0 I don't think this is testing anything of the syncer, or at least I can't work out what it is really testing. Maybe it's testing `Store.Load`. I would like to delete it. After looking at the new syncer_test.go, can you either 👍 or describe how to set up the case to be exercised without the shared setup of this file?

Reply:
This is an integration test I added after getting burned by bad assumptions between the store and the syncer taking nodes down.  Read [here](https://github.com/filecoin-project/go-filecoin/issues/1148#issuecomment-432008060) for a full description of what was being tested.

This particular edge case hasn't caused problems since this test was written and imo it is low risk to remove especially now that the chain store's block reading methods are gone.  

IMO we need to have integration tests that can hit specific cases like this without mocking out the chain store.  Ideally we move them to FAST.  I think this one could be ported to FAST without hassle.  `TipSetWeightDeep` would require more work but I think that it also belongs as a FAST integration test.  Having further syncer integration tests across a variety of edge cases is probably worthwhile, but may require significant CLI additions (i.e. a `sync start`/`sync once` subcommand) to do right.

Reply:
Thanks for the link. I'm not convinced FAST is the answer - we should be able to more directly put the store into some state then test the syncer deals with it. I think I understand the case now and will attempt to construct it in a unit test in a follow up.

*Specific* cases should be unit testable. Integration tests are good for the general case.